### PR TITLE
Expand tokens

### LINKED ISSUE #16: Expand tokens
Expand tokens based on posix arithmetic expansion.
Remplace dollars with env value.

### PR DESCRIPTION
**Pull request by Saphirac**
Expand tokens works. (?)
No leaks.
Normed.

### CODE REVIEW

File: src/env/env_lst/env_new.c [hunk 1]
```diff
@@ -27,7 +27,10 @@ t_env	*env_new(char const *const name, char const *const value)
 	if (!node)
 		return (NULL);
 	node->name = ft_strdup(name);
-	(value && (node->value = ft_strdup(value))) || (node->value = NULL);
+	if (value)
+		node->value = ft_strdup(value);
+	else
+		node->value = NULL;
```
Comment (JonathanDUFOUR) [inline src/env/env_lst/env_new.c]: Why this change? (again)

File: src/builtin/export/process_one.c [hunk 2]
```diff
@@ -98,7 +98,10 @@ int	process_one(t_env_lst *const env, char const *const str)
 			return (internal_error("export"));
 		return (EXIT_SUCCESS);
 	}
-	(*ptr && ++ptr) || (ptr = NULL);
+	if (*ptr)
+		++ptr;
+	else
+		ptr = NULL;
```
Comment (JonathanDUFOUR) [inline src/builtin/export/process_one.c]: Why this change? The previous version was already functional. Encountered errors were caused by the compiler `gcc`, but now we use `clang`, so we do not have errors.

File: src/main.c [hunk 3]
```diff
@@ -58,9 +58,6 @@ inline static int	__init_env(t_env_lst *const env, char const *const *ep)
 	return (EXIT_SUCCESS);
 }
 
-/**
- * @brief	Free the allocated memory of the shell context.
- */
```
Comment (JonathanDUFOUR) [inline src/main.c]: Why did you remove this comment?

File: src/tokens/append_to_ret.c [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+static inline char	*__leak_free_join(char const *const str0,
+		char const *const str1)
+{
+	char	*result;
+
+	result = ft_strjoin(str0, str1);
+	return (free((void *)str0), result);
+}
+
+// TODO : Remake the ret = ft_strjoin to have no leaks.
```
Comment (JonathanDUFOUR) [inline src/tokens/append_to_ret.c]: Is this `TODO` still up to date? If it is not, you should remove it.

File: src/tokens/expand_tokens.c [hunk 5]
```diff
@@ -3,112 +3,62 @@
...
+/*   Updated: 2023/03/27 20:20:43 by mcourtoi         ###   ########.fr       */
 /*                                                                            */
 /* ************************************************************************** */
 
-/*
-
 #include "minishell.h"
 
-char	*expand_single_quotes(char *token)
+void	print_str_lst(t_str_lst *str)
```
Comment (JonathanDUFOUR) [inline src/tokens/expand_tokens.c]: Is this function still needed or was it a debugging function? If so, you should remove it before merging to `main`

File: src/tokens/expand_tokens.c [hunk 6]
```diff
@@ -3,112 +3,62 @@
...
-		if (!ret)
-			return (NULL);
-		return (ret);
+		if (__need_to_expand(tmp->str) == true)
+		{
+			if (create_str_lst(tmp, &str) == EXIT_FAILURE)
+				return (EXIT_FAILURE);
+			if (expand_dollars_str_lst(env_lst, &str, tmp) == EXIT_FAILURE)
+				return (str_lst_clear(&str), EXIT_FAILURE);
+			print_str_lst(&str);
```
Comment (JonathanDUFOUR) [inline src/tokens/expand_tokens.c]: Just as said before, if this was a debugging function, you should remove it before merging into `main`.

File: src/tokens/lst_tokens/token_lst_add_after.c [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+	t_token	*tmp;
+
+	if (!token->next)
+		return (token_lst_add_back(token_lst, type, str));
+	node = token_new(type, str);
+	if (!node)
+		return (EXIT_FAILURE);
+	tmp = token->next;
+	token->next = node;
+	while (tmp)
```
Comment (JonathanDUFOUR) [inline src/tokens/lst_tokens/token_lst_add_after.c]: It is useless to relink all the nodes placed after the one you just inserted. Only the first node placed after the one you inserted is to be relinked. Not the others.

File: src/tokens/search_env.c [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+(non alphanumeric character and non '_') 
+Check if name between that corresponds to an env value 
+If true, copy the value 
+else return a NULL chain */
+
+int	append_if_exit_code(char **ret, t_str *str, t_tmp_i_start *i)
+{
+	char	*tmp;
+
+	tmp = ft_itoa(g_exit_code);
```
Comment (JonathanDUFOUR) [inline src/tokens/search_env.c]: `g_exit_code` has type `uint8_t`. Therefore, the function `ft_hhutoa()` would be more relevant than `ft_itoa()`.

File: src/tokens/split_spaces.c [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+/*   Updated: 2023/03/27 21:06:38 by mcourtoi         ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "minishell.h"
+
+int	split_is_null(t_token_lst *const token_lst,
+		t_token **const token, t_str const *const tmp)
+{
+	if (ft_strlen((*token)->str))
```
Comment (JonathanDUFOUR) [inline src/tokens/split_spaces.c]: I assume that you are checking whether `(*token)->str` is an empty string. If so, just checking `*(*token)->str` instead of calling `ft_strlen()` would suffice, and it always takes  1 instruction instead of N.

File: src/tokens/split_spaces.c [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+			(*token) = (*token)->next;
+		}
+	}			
+	return (EXIT_SUCCESS);
+}
+
+int	add_spaces(t_token_lst *const token_lst, t_token **token,
+		t_str *tmp, char **split)
+{
+	if (ft_strlen((*token)->str))
```
Comment (JonathanDUFOUR) [inline src/tokens/split_spaces.c]: Same reflection, if you only intend to check whether `(*token)->str` is an empty string, only checking `*(*token)->str` would suffice.

File: src/tokens/split_utils.c [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+{
+	return (is_white_spaces(str[0])
+		|| is_white_spaces(str[ft_strlen(str) - 1]));
+}
+
+int	strdup_or_join(t_token *const *const token, char const *const tmp)
+{
+	char	*tmp_ptr;
+
+	if (ft_strlen((*token)->str) == 0)
```
Comment (JonathanDUFOUR) [inline src/tokens/split_utils.c:30]: (Again) You should check only the first character instead of calculating the whole length.

File: src/tokens/str_lst/str_new.c [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+t_str	*str_new(char *const str, bool is_quoted)
+{
+	t_str *const	output = malloc(sizeof(t_str));
+
+	if (!output)
+		return (NULL);
+	output->str = ft_strdup(str);
+	if (!output->str)
+		return (NULL);
```
Comment (JonathanDUFOUR) [inline src/tokens/str_lst/str_new.c]: If this condition is true (if `ft_strdup()` fails), then you forgot to `free(output)`, and thus you have a definitely lost memory leak.

**Review by JonathanDUFOUR — CHANGES_REQUESTED** [review]
Several minor changes have to be done before the pull request would be approved. Except for these little changes to do, no important functional mistakes have been made. The last time I tested it, it worked well! Good job!

File: src/env/env_lst/env_new.c (hunk 1 above)
Comment (Saphirac) [reply src/env/env_lst/env_new.c]: Compilator probably didn't work.

File: src/builtin/export/process_one.c (hunk 2 above)
Comment (Saphirac) [reply src/builtin/export/process_one.c]: It still works so I think we can keep it.

File: src/main.c (hunk 3 above)
Comment (Saphirac) [reply src/main.c]: I don't think I removed it on purpose. (Maybe you did?)

File: src/tokens/append_to_ret.c (hunk 4 above)
Comment (Saphirac) [reply src/tokens/append_to_ret.c]: Todo is good now.

File: src/tokens/expand_tokens.c (hunk 5 above)
Comment (Saphirac) [reply src/tokens/expand_tokens.c]: Function not needed anymore will delete.

File: src/tokens/expand_tokens.c (hunk 6 above)
Comment (Saphirac) [reply src/tokens/expand_tokens.c]: Will be deleted.

File: src/tokens/lst_tokens/token_lst_add_after.c (hunk 7 above)
Comment (Saphirac) [reply src/tokens/lst_tokens/token_lst_add_after.c]: What?

File: src/tokens/search_env.c (hunk 8 above)
Comment (Saphirac) [reply src/tokens/search_env.c]: Noted.

File: src/tokens/split_spaces.c (hunk 9 above)
Comment (Saphirac) [reply src/tokens/split_spaces.c]: No. (*token)->str is strdup(""); so it won't works to just check whether it exists or not.

File: src/tokens/str_lst/str_new.c (hunk 12 above)
Comment (Saphirac) [reply src/tokens/str_lst/str_new.c]: It is true.

**Review by JonathanDUFOUR — APPROVED** [review]
Perfect, every requested change has been well considered and made. We can merge it on the `main` branch now.
